File: prog_tut_sample_solutions/java/starter/Bank.java
/**
COMP3331/9331 Computer Networks and Applications
Programming Tutorial - Online Banking System
Written by Gary Hu

This file contains the java bank class for the online banking system.
 */
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;

public class Bank {
    private Map<String, Account> accounts;

    /**
     * Initialise the bank, using the account information provided in the accountsFile.
     * The file should contain one account per line, with the account name, hashed password,
     * and balance separated by whitespace.
     *
     * @param accountsFile Path to the file containing account information.
     */
    public Bank(String accountsFile) {
        this.accounts = loadAccounts(accountsFile);
    }

    /**
     * Adds an account to the bank database. The account name must be unique. The opening balance is zero.
     *
     * @param account      A name to unique identify the account.
     * @param passwordHash A SHA-1 hash of the account password.
     * @return A response phrase, indicating success or failure.
     */
    public String openAccount(String account, String passwordHash) {
        if (accounts.containsKey(account)) {
            return "account already exists";
        }
        accounts.put(account, new Account(passwordHash, 0.0));
        return "successful";
    }

    /**
     * Returns the balance of the account. The request is only successful if the account exists
     * and the provided password hash matches the database.
     *
     * @param account      Account name.
     * @param passwordHash SHA-1 hash of the account password.
     * @return The balance, formatted as a string with two decimal places.
     */
    public String getBalance(String account, String passwordHash) {
        if (!isAuthorized(account, passwordHash)) {
            return "not authorised";
        }
        return String.format("%.2f", accounts.get(account).getBalance());
    }

    /**
     * Transfer funds from one account to another. The transfer is only successful if the fromAccount exists,
     * the hashed password matches, the toAccount exists, and the fromAccount has sufficient funds.
     *
     * @param fromAccount  Account name of the sender.
     * @param passwordHash SHA-1 password hash of the sender.
     * @param toAccount    Account name of the receiver.
     * @param amount       Amount to transfer.
     * @return A response phrase, indicating success or failure.
     */
    public String transferFunds(String fromAccount, String passwordHash, String toAccount, double amount) {
        if (!isAuthorized(fromAccount, passwordHash) || !accountExists(toAccount)) {
            return "not authorised";
        }

        // TODO 8: Critical section of code, ensure that only one thread can
        // access the bank accounts at a time.
    }

    /**
     * Check if the account exists in the bank database.
     *
     * @param account Account name.
     * @return True if the account exists, false otherwise.
     */
    private boolean accountExists(String account) {
        return accounts.containsKey(account);
    }

    /**
     * Check if the account exists and the provided password hash matches.
     *
     * @param account      Account name.
     * @param passwordHash SHA-1 hash of the account password.
     * @return True if authorized, false otherwise.
     */
    private boolean isAuthorized(String account, String passwordHash) {
        if (!accountExists(account)) {
            return false;
        }
        return accounts.get(account).getPasswordHash().equals(passwordHash);
    }

    /**
     * Check if the account has sufficient funds to transfer the amount.
     *
     * @param account Account name.
     * @param amount  Amount to transfer.
     * @return True if sufficient funds, false otherwise.
     */
    private boolean hasSufficientBalance(String account, double amount) {
        return accounts.get(account).getBalance() >= amount;
    }

    /**
     * Update the balance of the account by the specified amount.
     *
     * @param account Account name.
     * @param amount  Amount to update.
     */
    private void updateBalance(String account, double amount) {
        Account acc = accounts.get(account);
        acc.setBalance(acc.getBalance() + amount);
    }

    /**
     * Load the account information from a file.
     *
     * @param filename Path to the file containing account information.
     * @return A map of account names to Account objects.
     */
    private static Map<String, Account> loadAccounts(String filename) {
        Map<String, Account> accounts = new HashMap<>();
        Path filepath = Paths.get(filename);

        if (!Files.exists(filepath)) {
            System.err.println("Error: " + filename + " does not exist.");
            System.exit(1);
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split("\\s+");
                String account = parts[0];
                String passwordHash = parts[1];
                double balance = Double.parseDouble(parts[2]);
                accounts.put(account, new Account(passwordHash, balance));
            }
        } catch (IOException e) {
            e.printStackTrace();
            System.exit(1);
        }

        return accounts;
    }
}

File: prog_tut_sample_solutions/python/starter/bank.py
"""
COMP3331/9331 Computer Networks and Applications
Programming Tutorial - Online Banking System
Written by Tim Arney <t.arney@unsw.edu.au>

This file contains the bank class for the online banking system.
"""

from dataclasses import dataclass
from pathlib import Path
import sys
import threading


@dataclass
class Account:
    """Bank will store a dict of accounts, where the key is the account name"""
    password_hash: str
    balance: float = 0.0


class Bank:
    def __init__(self, accounts_file: str):
        """Initialise the bank, using the account information provided in the
           accounts_file.  The file should contain one account per line, with
           the account name, hashed password, and balance separated by 
           whitespace.

        Args:
            accounts_file (str): Path to the file containing account information.
        """
        self.accounts = self._load_accounts(accounts_file)

        # TODO 8.1: Add a lock primitive, assign it to a member variable of the class.


    def open_account(self, account: str, password_hash: str) -> str:
        """Adds an account to the bank database.  The account name must be 
           unique.  The opening balance is zero.

        Args:
            account (str): A name to unique identify the account.
            password_hash (str): A SHA-1 hash of the account password.

        Returns:
            str: A response phrase, indicating success or failure.
        """
        if account in self.accounts:
            return "account already exists"

        self.accounts[account] = Account(password_hash)
        return "successful"

    def get_balance(self, account: str, password_hash: str) -> str:
        """Retuns the balance of the account.  The requests is only successful 
           if the account exists and the provided password hash matches the 
           database.

        Args:
            account (str): Account name.
            password_hash (str): SHA-1 hash of the account password.

        Returns:
            str: The balance, formatted as a string with two decimal places.
        """
        if not self._is_authorised(account, password_hash):
            return "not authorised"
        
        return f"{self.accounts[account].balance:.2f}"

    def transfer_funds(self, from_account: str, password_hash: str, to_account: str, amount: float) -> str:
        """Transfer funds from one account to another.  The transfer is only 
              successful if the from_account exists, the hashed password matches, 
              the to_account exists, and the from_account has sufficient funds.

        Args:
            from_account (str): Account name of the sender.
            password_hash (str): SHA-1 password hash of the sender.
            to_account (str): Account name of the receiver.
            amount (float): Amount to transfer.

        Returns:
            str: A response phrase, indicating success or failure.
        """
        if not self._is_authorised(from_account, password_hash) or not self._account_exists(to_account):
            return "not authorised"
        
        # TODO 8.2: Critical section of code, ensure that only one thread can
        #   access the bank accounts at a time.
        if not self._has_sufficient_balance(from_account, amount):
            return "insufficient funds"

        self._update_balance(from_account, -amount)
        self._update_balance(to_account, amount)
        # end of critical section

        return "successful"

    def _account_exists(self, account: str) -> bool:
        """Check if the account exists in the bank database."""
        return account in self.accounts
    
    def _is_authorised(self, account: str, password_hash: str) -> bool:
        """Check if the account exists and the provided password hash matches."""
        if not self._account_exists(account):
            return False

        return self.accounts[account].password_hash == password_hash

    def _has_sufficient_balance(self, account: str, amount: str) -> bool:
        """Check if the account has sufficient funds to transfer the amount."""
        return self.accounts[account].balance >= amount

    def _update_balance(self, account: str, amount: float) -> None:
        """Update the balance of the account by the specified amount."""
        self.accounts[account].balance += amount
    
    @staticmethod
    def _load_accounts(filename: str) -> dict[str, Account]:
        """Load the account information from a file."""
        accounts = {}
        filepath = Path(filename)
        
        if not filepath.exists():
            sys.exit(f'Error: {filename} does not exist.')

        with open(filename, 'r') as f:
            for line in f:
                account, password_hash, balance = line.split()
                accounts[account] = Account(password_hash, float(balance))

        return accounts
